common/js/src/messaging/message-channel-pair-unittest.js

```diff
@@ -0,0 +1,113 @@
+// the second item of the pair.
+goog.exportSymbol('test_MessageChannelPair_sendViaFirst', function() {
+  return new Promise((resolve, reject) => {
+    const messageChannelPair = new GSC.MessageChannelPair();
+
+    // Set up the expectation - once the right message is received on the
+    // second item, make the test complete by resolving the promise.
+    messageChannelPair.getSecond().registerService(
+        /*serviceName=*/ 'some-service', (payload) => {
+          assertEquals(payload, 'data');
+          resolve();
+        });
+    // Any unexpected messages on the second item should abort the test.
+    messageChannelPair.getSecond().registerDefaultService((payload) => {
+      reject();
+    })
+
+    messageChannelPair.getFirst().send(
+        /*serviceName=*/ 'some-service', /*payload=*/ 'data');
+  });
+});
+
+// Inversion of the test above.
+goog.exportSymbol('test_MessageChannelPair_sendViaSecond', function() {
+  return new Promise((resolve, reject) => {
+    const messageChannelPair = new GSC.MessageChannelPair();
+
+    messageChannelPair.getFirst().registerService(
+        /*serviceName=*/ 'some-service', (payload) => {
+          assertEquals(payload, 'data');
+          resolve();
+        });
+    messageChannelPair.getFirst().registerDefaultService((payload) => {
+      reject();
+    })
+
+    messageChannelPair.getSecond().send(
+        /*serviceName=*/ 'some-service', /*payload=*/ 'data');
+  });
+});
+
+// That that the disposal of the channel pair causes disposing of both items of
+// the pair.
+goog.exportSymbol('test_MessageChannelPair_dispose', function() {
+  const messageChannelPair = new GSC.MessageChannelPair();
+  const first = messageChannelPair.getFirst();
+  const second = messageChannelPair.getSecond();
+  messageChannelPair.dispose();
+  assertTrue(first.isDisposed());
+  assertTrue(second.isDisposed());
+});
+
+// That that the disposal of one item of the channel pair causes disposing of
+// the channel pair itself and the other item too.
+goog.exportSymbol('test_MessageChannelPair_disposeItem', function() {
+  const messageChannelPair = new GSC.MessageChannelPair();
+  const first = messageChannelPair.getFirst();
+  const second = messageChannelPair.getSecond();
+  first.dispose();
+  assertTrue(messageChannelPair.isDisposed());
+  assertTrue(second.isDisposed());
+});
+
+// Test that the message to be sent is discarded if the channel pair is disposed
+// of before the sending task starts.
+goog.exportSymbol(
+    'test_MessageChannelPair_disposePairQuicklyAfterSend', function() {
+      return new Promise((resolve, reject) => {
+        const messageChannelPair = new GSC.MessageChannelPair();
+
+        // Abort the test if the message got delivered despite the disposal.
+        messageChannelPair.getSecond().registerDefaultService((payload) => {
+          reject();
+        })
+
+        messageChannelPair.getFirst().send(
+            /*serviceName=*/ 'some-service', /*payload=*/ 'data');
```

Comment:
Is there any way that this may complete before the message channel is disposed?

Reply:
Good question - It can't happen in the current code, but I've added comments to explain that.

Please reopen if you think the comments are unclear or the behavior needs to be changed.